rspace/src/main/scala/coop/rchain/rspace/history/HistoryRepositoryImpl.scala

```diff
@@ -183,22 +184,24 @@ final case class HistoryRepositoryImpl[F[_]: Sync, C, P, A, K](
         (key, None, DeleteAction(key.bytes.toSeq.toList))
     }
 
-  private def storeLeaves(leafs: List[Result]): F[List[HistoryAction]] =
-    leafs.traverse {
-      case (key, Some(data), historyAction) =>
-        leafStore.put(key, data).map(_ => historyAction)
-      case (_, None, historyAction) =>
-        Applicative[F].pure(historyAction)
-    }
+  private def storeLeaves(leafs: List[Result]): F[List[HistoryAction]] = {
+    val toBeStored = leafs.collect { case (key, Some(data), _) => (key, data) }
+    leafStore.put(toBeStored).map(_ => leafs.map(_._3))
+  }
 
-  override def checkpoint(actions: List[HotStoreAction]): F[HistoryRepository[F, C, P, A, K]] =
+  override def checkpoint(actions: List[HotStoreAction]): F[HistoryRepository[F, C, P, A, K]] = {
+    implicit val P = Par[F].parallel
+    val batchSize  = Math.max(1, actions.size / RSpace.parallelism)
     for {
-      trieActions    <- Applicative[F].pure(actions.par.map(transform).toList)
-      historyActions <- storeLeaves(trieActions)
-      next           <- history.process(historyActions)
-      _              <- rootsRepository.commit(next.root)
-      _              <- measure(actions)
+      batches               <- Sync[F].delay(actions.grouped(batchSize).toList)
+      trieActionsBatched    <- batches.parTraverse(b => Sync[F].delay(b.map(transform)))
+      historyActionsBatched <- trieActionsBatched.parTraverse(storeLeaves)
```

Comment:
I think we can avoid the Sync.delay by doing:
`trieActionsBatched.parTraverse(transform.andThen(storeLeaves))`
or sth similar?